### PR TITLE
Merkle flag

### PR DESCRIPTION
**Pull request by kevinssgh**
Added a flag to determine whether the backend uses RocksDB or MerkDB

* The range checks on the key and values will be ignored for RocksDB and enforced for MerkDB

### CODE REVIEW

File: src/state/mod.rs [hunk 1]
```diff
@@ -387,9 +387,49 @@ mod tests {
...
+        // Setup
+        let path = thread::current().name().unwrap().to_owned();
+        let cs = gen_cs_rocks(path);
+
+        // Make sure is_merkle flag is false
+        let mut state = State::new(cs, false);
+
+        // Set a big value
+        let big_val = "v".repeat(u16::MAX as usize + 1).as_bytes().to_vec();
+        assert!(state.set(b"key10", big_val).is_ok());
```
Comment (ws4charlie) [inline src/state/mod.rs:401]: Looks great to me. BTW, can you also please remove the println!("") on BIG value and print only the key when a KV fails to pass the checking?

Comment (kevinssgh) [reply src/state/mod.rs:401]: You mean this part:

        // check value
        if value.len() > MAX_MERK_VAL_LEN as usize {
            let val_str =
                String::from_utf8(value.to_vec()).map_or("non-utf8-value".to_owned(), |v| v);
            println!("Invalid value length: {}", value.len());
            return false;
        }

Comment (kevinssgh) [reply src/state/mod.rs:401]: I changed it to:

 // check value
        if value.len() > MAX_MERK_VAL_LEN as usize {
            println!("Invalid value length: {}", value.len());
            return false;
        }

Comment (ws4charlie) [reply src/state/mod.rs:401]: Looks good to me. Thanks

**Review by ws4charlie — APPROVED** [review]

**Pull request merged**